packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -544,22 +608,67 @@ class SliverChildListDelegate extends SliverChildDelegate {
   /// The widgets to display.
   final List<Widget> children;
 
+  final Iterator<Widget> _childIterator;
+
+  final Map<Key, int> _mapKeyToIndex;
+
+  final bool _isConstantInstance;
+
+  int _findItemIndex(Key key){
+    if(_isConstantInstance) {
+      return -1;
+    }
+    // This function should only be called when any child has associated key.
+    // We do lazy loading in this function.
+    if(!_mapKeyToIndex.containsKey(key)) {
+      int index;
+      while(_childIterator.moveNext()) {
+        final Widget child = _childIterator.current;
+        index ??= children.indexOf(child);
+        if (child.key != null) {
+          _mapKeyToIndex[child.key] = index;
+        }
+        if (child.key == key) {
+          break;
+        }
+        index++;
+      }
+    }
+    return _mapKeyToIndex.containsKey(key) ? _mapKeyToIndex[key] : -1;
+  }
+
+  @override
+  int findIndexByKey(Key key) {
+    assert(key != null);
+    Key childKey;
+    if (addAutomaticKeepAlives || addRepaintBoundaries || addSemanticIndexes) {
+      // in this case, child key is wrapped with ValueKey<Key>
+      final ValueKey<Key> valueKey = key;
+      assert(valueKey.value != null);
+      childKey = valueKey.value;
+    } else {
+      childKey = key;
+    }
+    return _findItemIndex(childKey);
```

Comment:
What if childKey is null? Should we just return -1 directly?

Reply:
if childKey is null, the SliverChildListDelegator will not wrap the key at the first place, and this function should not get called line 647 is also an assert to sanitize it